kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,305 @@
+    exp: &Expr<C>,
+    flag_resolver: &F,
+) -> IntegratedFoldingExpr<C> {
+    let simplied = simplify_expression(exp, flag_resolver);
+    let mut terms = vec![];
+    extract_terms(simplied, &mut terms, &|t| t);
+    let mut integrated = IntegratedFoldingExpr::default();
+    for term in terms.into_iter() {
+        let Term { exp, sign, degree } = term;
+        let t = (exp, sign);
+        match degree {
+            Degree::Zero => integrated.degree_0.push(t),
+            Degree::One => integrated.degree_1.push(t),
+            Degree::Two => integrated.degree_2.push(t),
+        }
+    }
+    integrated
+}
+fn simplify_expression<C: Copy, F: Fn(&FeatureFlag) -> bool>(
+    exp: &Expr<C>,
+    flag_resolver: &F,
+) -> FoldingExp<C> {
+    match exp {
+        Expr::Constant(c) => FoldingExp::Constant(*c),
+        Expr::Cell(v) => {
+            let Variable { col, row } = v;
+            let col = match col {
+                Column::Witness(i) => FoldingColumn::Witness(*i),
+                Column::Index(i) => FoldingColumn::Index(*i),
+                Column::Coefficient(i) => FoldingColumn::Coefficient(*i),
+                _ => unreachable!(),
+            };
+            let cell = Box::new(FoldingExp::Cell(col));
+            let shift = Box::new(FoldingExp::Cell(FoldingColumn::Shift));
+            match row {
+                CurrOrNext::Curr => *cell,
+                CurrOrNext::Next => FoldingExp::Mul(cell, shift),
+            }
+        }
+        //may be better to replace by mul by 2
+        Expr::Double(e) => {
+            let e = simplify_expression(e, flag_resolver);
+            FoldingExp::Double(Box::new(e))
+        }
+        //may be better to replace by mul by itself
+        Expr::Square(e) => {
+            let e = simplify_expression(e, flag_resolver);
+            FoldingExp::Square(Box::new(e))
+        }
+        Expr::BinOp(op, e1, e2) => {
+            let e1 = Box::new(simplify_expression(e1, flag_resolver));
+            let e2 = Box::new(simplify_expression(e2, flag_resolver));
+            match op {
+                Op2::Add => FoldingExp::Add(e1, e2),
+                Op2::Sub => FoldingExp::Sub(e1, e2),
+                Op2::Mul => FoldingExp::Mul(e1, e2),
+            }
+        }
+        Expr::UnnormalizedLagrangeBasis(i) => {
+            assert!(!i.zk_rows);
+            FoldingExp::UnnormalizedLagrangeBasis(i.offset)
+        }
+        Expr::Cache(id, e) => {
+            let e = simplify_expression(e, flag_resolver);
+            FoldingExp::Cache(*id, Box::new(e))
+        }
+        Expr::IfFeature(flag, e1, e2) => {
+            if flag_resolver(flag) {
+                simplify_expression(e1, flag_resolver)
+            } else {
+                simplify_expression(e2, flag_resolver)
+            }
+        }
+        Expr::Pow(_, _) => {
+            unreachable!()
+        }
+        //TODO: check that this is fine
+        Expr::VanishesOnZeroKnowledgeAndPreviousRows => todo!(),
```

Comment:
What's a semantics of this even?

Reply:
I'm not completely sure right now, it was added after I started this branch, I believe that it is 0 in the las 4 rows used for ZK, but it replaced another variant that already did that making me think that it does something more.
It is low priority now given that we won't be using ZK with folding for now at least.